Fiat API

xmrdsc opened the pull request:
![](https://i.imgur.com/VUbDVpH.png)

xiphon commented on src/qt/utils.cpp:
```diff
@@ -18,3 +18,84 @@ QString getAccountName(){
+          << "Mozilla/5.0 (Windows NT 6.1; de;rv:12.0) Gecko/20120403211507 Firefox/12.0";
+
+    // @TODO: Qt 5.10 - QRandomNumberGenerator
+    int irand = (rand() % urand.length()) + 1;
```
```suggestion
    int irand = rand() % urand.length();
```

xmrdsc replied:
:+1:

SamsungGalaxyPlayer commented:
What is the point of the "price conversion" box?

xmrdsc commented:
Conversion for arbitrary amounts; convenience.

xiphon commented:
Would be better to always display both XMR and fiat balances when fiat conversion feature is enabled.
And, yep, think we don't need "Price conversion" box in the settings. It is almost useless.

selsta commented:
`Settings -> Interface` is a weird place for a price conversion tool. I’d say such a tool doesn’t fit into the GUI.

> Would be better to always display both XMR and fiat balances when fiat conversion feature is enabled.

I like the current way, fiat only when hovering the balance card.

selsta commented:
```
2019-05-01 18:03:19.369	W qrc:/pages/settings/SettingsLayout.qml:304: TypeError: Cannot call method 'start' of undefined
2019-05-01 18:04:46.723	W qrc:/pages/settings/SettingsLayout.qml:171: TypeError: Cannot call method 'stop' of undefined
2019-05-01 18:04:58.257	W qrc:/pages/settings/SettingsLayout.qml:171: TypeError: Cannot call method 'stop' of undefined
```

Some more feedback:

- The extra enable price conversion button is confusing.
- Like I wrote above, I’d remove the conversion box.
- Proper currency symbols are missing (€, $, ...)
- Copying balance isn’t working after this change.

xmrdsc commented:
Updated;

- Removed conversion tool (booo :P)
- Fixed undefined references
- Added currency symbols
- Copying balance also does not work for me on selsta's fiat PR

Thanks for reviews.

mmbyday commented:
> * Removed conversion tool (booo :P)

FWIW, this tool would be valuable if built-in to the transfer page. Make it as easy as sending xmr, or some value denominated in fiat.

selsta commented:
```diff
diff --git a/LeftPanel.qml b/LeftPanel.qml
index 846783c..c1e907c 100644
--- a/LeftPanel.qml
+++ b/LeftPanel.qml
@@ -32,6 +32,7 @@ import QtGraphicalEffects 1.0
 import moneroComponents.Wallet 1.0
 import moneroComponents.NetworkType 1.0
 import moneroComponents.Clipboard 1.0
+import FontAwesome 1.0
 
 import "components" as MoneroComponents
 import "components/effects/" as MoneroEffects
@@ -50,6 +51,7 @@ Rectangle {
     property alias progressBar : progressBar
     property alias daemonProgressBar : daemonProgressBar
     property alias minutesToUnlockTxt: unlockedBalanceLabel.text
+    property bool fiatBalance: false
     property int titleBarHeight: 50
     property string copyValue: ""
     Clipboard { id: clipboard }
@@ -115,12 +117,6 @@ Rectangle {
         anchors.top: parent.top
         anchors.topMargin: (persistentSettings.customDecorations)? 50 : 0
 
-        MouseArea {
-            id: balanceArea
-            anchors.fill: parent
-            hoverEnabled: true
-        }
-
         RowLayout {
             Item {
                 anchors.left: parent.left
@@ -210,6 +206,24 @@ Rectangle {
                         }
                     }
                 }
+                MoneroComponents.Label {
+                    fontSize: 20
+                    text: "¥"
+                    visible: persistentSettings.fiatPriceEnabled
+                    anchors.right: parent.right
+                    anchors.rightMargin: 45
+                    anchors.top: parent.top
+                    anchors.topMargin: 29
+
+                    MouseArea{
+                        anchors.fill: parent
+                        hoverEnabled: true
+                        cursorShape: Qt.PointingHandCursor
+                        onClicked: {
+                            fiatBalance = !fiatBalance
+                        }
+                    }
+                }
             }
 
             Item {
@@ -222,7 +236,7 @@ Rectangle {
                 width: 50
 
                 MoneroComponents.TextPlain {
-                    visible: !(balanceArea.containsMouse && persistentSettings.fiatPriceEnabled)
+                    visible: !(fiatBalance && persistentSettings.fiatPriceEnabled)
                     id: balanceText
                     themeTransition: false
                     anchors.left: parent.left
@@ -275,12 +289,28 @@ Rectangle {
                     color: "#FFFFFF"
                     text: "N/A"
                     font.pixelSize: balanceText.font.pixelSize
+                    MouseArea {
+                        hoverEnabled: true
+                        anchors.fill: parent
+                        cursorShape: Qt.PointingHandCursor
+                        onEntered: {
+                            parent.color = MoneroComponents.Style.orange
+                        }
+                        onExited: {
+                            parent.color = MoneroComponents.Style.white
+                        }
+                        onClicked: {
+                                console.log("Copied to clipboard");
+                                clipboard.setText(parent.text);
+                                appWindow.showStatusMessage(qsTr("Copied to clipboard"),3)
+                        }
+                    }
                 }
 
                 MoneroComponents.TextPlain {
                     id: unlockedBalanceText
 
-                    visible: !(balanceArea.containsMouse && persistentSettings.fiatPriceEnabled)
+                    visible: !(fiatBalance && persistentSettings.fiatPriceEnabled)
                     themeTransition: false
                     anchors.left: parent.left
                     anchors.leftMargin: 20
@@ -332,6 +362,22 @@ Rectangle {
                     color: "#FFFFFF"
                     text: "N/A"
                     font.pixelSize: unlockedBalanceText.font.pixelSize
+                    MouseArea {
+                        hoverEnabled: true
+                        anchors.fill: parent
+                        cursorShape: Qt.PointingHandCursor
+                        onEntered: {
+                            parent.color = MoneroComponents.Style.orange
+                        }
+                        onExited: {
+                            parent.color = MoneroComponents.Style.white
+                        }
+                        onClicked: {
+                                console.log("Copied to clipboard");
+                                clipboard.setText(parent.text);
+                                appWindow.showStatusMessage(qsTr("Copied to clipboard"),3)
+                        }
+                    }
                 }
 
                 MoneroComponents.Label {
diff --git a/main.qml b/main.qml
index c94dc09..f023d5b 100644
--- a/main.qml
+++ b/main.qml
@@ -415,7 +415,9 @@ ApplicationWindow {
         middlePanel.balanceText = balance;
         leftPanel.balanceText = balance;
 
-        appWindow.fiatApiUpdateBalance(balance, balance_unlocked);
+        if (persistentSettings.fiatPriceEnabled) {
+            appWindow.fiatApiUpdateBalance(balance, balance_unlocked);
+        }
 
         var accountLabel = currentWallet.getSubaddressLabel(currentWallet.currentSubaddressAccount, 0);
         leftPanel.balanceLabelText = qsTr("Balance (#%1%2)").arg(currentWallet.currentSubaddressAccount).arg(accountLabel === "" ? "" : (" – " + accountLabel));
diff --git a/pages/settings/SettingsLayout.qml b/pages/settings/SettingsLayout.qml
index 2444f9a..c211cc4 100644
--- a/pages/settings/SettingsLayout.qml
+++ b/pages/settings/SettingsLayout.qml
@@ -178,7 +178,7 @@ Rectangle {
             visible: enableConvertCurrency.checked
             columns: 2
             Layout.fillWidth: true
-            Layout.topMargin: 12
+            Layout.leftMargin: 36
             columnSpacing: 32
 
             ColumnLayout {
@@ -232,18 +232,29 @@ Rectangle {
             z: parent.z + 1
         }
 
-        MoneroComponents.StandardButton {
+        ColumnLayout {
             // Feature needs to be double enabled for security purposes (miss-clicks)
             visible: persistentSettings.fiatPriceEnabledCheckbox && !persistentSettings.fiatPriceEnabled
-            Layout.topMargin: 10
-            small: true
-            text: qsTr("Enable price conversion") + translationManager.emptyString
+            spacing: 0
+            Layout.topMargin: 5
+            Layout.leftMargin: 36
 
-            onClicked: {
-                console.log("Enabled price conversion");
-                persistentSettings.fiatPriceEnabled = true;
-                appWindow.fiatApiRefresh();
-                appWindow.fiatTimerStart();
+            MoneroComponents.WarningBox {
+                text: qsTr("Enabling price conversion exposes your IP address to the selected price source.") + translationManager.emptyString;
+            }
+
+            MoneroComponents.StandardButton {
+                Layout.topMargin: 10
+                Layout.bottomMargin: 10
+                small: true
+                text: qsTr("Confirm and enable") + translationManager.emptyString
+
+                onClicked: {
+                    console.log("Enabled price conversion");
+                    persistentSettings.fiatPriceEnabled = true;
+                    appWindow.fiatApiRefresh();
+                    appWindow.fiatTimerStart();
+                }
             }
         }
 
```

xmrdsc commented:
@mmbyday That would be nice. Will have to see how that could work with our design/UX.

selsta commented on LeftPanel.qml:
```diff
@@ -255,9 +277,40 @@ Rectangle {
                 MoneroComponents.TextPlain {
                     id: unlockedBalanceText
-                    visible: true
+
```
```suggestion

```

xmrdsc replied:
👍

selsta commented on main.qml:
```diff
@@ -1228,6 +1393,10 @@ ApplicationWindow {
         property bool showPid: false
         property bool blackTheme: false
+        property bool fiatPriceEnabled: false
+        property bool fiatPriceEnabledCheckbox: false
```
I don’t see a reason to save fiatPriceEnabledCheckbox here.

xmrdsc replied:
That is correct, I've set it locally (`appWindow`) instead. Thanks.

selsta commented on pages/settings/SettingsLayout.qml:
```diff
@@ -177,7 +278,43 @@ Rectangle {
+            i += 1;
+        }
+
+        fiatPriceProviderDropDown.update();
```
Is update twice intentional?

xmrdsc replied:
`fiatPriceProviderDropDown` && `fiatPriceCurrencyDropdown `

selsta commented:
```diff
diff --git a/main.qml b/main.qml
index afaf170..36ddf6d 100644
--- a/main.qml
+++ b/main.qml
@@ -93,7 +93,6 @@ ApplicationWindow {
     property bool themeTransition: false
 
     // fiat price conversion
-    property bool fiatPriceEnabledCheckbox: false
     property int fiatPriceXMRUSD: 0
     property int fiatPriceXMREUR: 0
     property var fiatPriceAPIs: {
@@ -1395,7 +1394,6 @@ ApplicationWindow {
         property bool showPid: false
         property bool blackTheme: false
         property bool fiatPriceEnabled: false
-        property bool fiatPriceEnabledCheckbox: false
         property string fiatPriceProvider: "kraken"
         property string fiatPriceCurrency: "xmrusd"
 
diff --git a/pages/settings/SettingsLayout.qml b/pages/settings/SettingsLayout.qml
index 6303bc7..b8dbcad 100644
--- a/pages/settings/SettingsLayout.qml
+++ b/pages/settings/SettingsLayout.qml
@@ -162,10 +162,9 @@ Rectangle {
             MoneroComponents.CheckBox {
                 id: enableConvertCurrency
                 text: qsTr("Enable displaying balance in other currencies") + translationManager.emptyString
-                checked: appWindow.fiatPriceEnabledCheckbox
+                checked: persistentSettings.fiatPriceEnabled
                 onCheckedChanged: {
-                    appWindow.fiatPriceEnabledCheckbox = checked;
-                    if(!checked){
+                    if (!checked) {
                         console.log("Disabled price conversion");
                         persistentSettings.fiatPriceEnabled = false;
                         appWindow.fiatTimerStop();
@@ -234,7 +233,7 @@ Rectangle {
 
         ColumnLayout {
             // Feature needs to be double enabled for security purposes (miss-clicks)
-            visible: appWindow.fiatPriceEnabledCheckbox && !persistentSettings.fiatPriceEnabled
+            visible: enableConvertCurrency.checked && !persistentSettings.fiatPriceEnabled
             spacing: 0
             Layout.topMargin: 5
             Layout.leftMargin: 36
```

xmrdsc commented:
True, added. Thanks @selsta

mmbyday commented:
Getting an error on local window builds and seems it happens on buildbot's window build too. Am I missing a QSslSocket declaration somewhere?
```
"Fetching: https://min-api.cryptocompare.com/data/price?fsym=XMR&tsyms=USD"
WARNING	frontend	src/wallet/api/wallet.cpp:410	QSslSocket: cannot call unresolved function SSLv23_client_method
WARNING	frontend	src/wallet/api/wallet.cpp:410	QSslSocket: cannot call unresolved function SSL_CTX_new
WARNING	frontend	src/wallet/api/wallet.cpp:410	QSslSocket: cannot call unresolved function SSL_library_init
WARNING	frontend	src/wallet/api/wallet.cpp:410	QSslSocket: cannot call unresolved function ERR_get_error
WARNING	frontend	src/wallet/api/wallet.cpp:410	QSslSocket: cannot call unresolved function ERR_get_error
ERROR	frontend	src/wallet/api/wallet.cpp:414	gotError : Error:  "Unknown error"
```

selsta commented:
```diff
diff --git a/LeftPanel.qml b/LeftPanel.qml
index a46459c..82d5b76 100644
--- a/LeftPanel.qml
+++ b/LeftPanel.qml
@@ -209,11 +209,13 @@ Rectangle {
                 MoneroComponents.Label {
                     fontSize: 20
                     text: "¥"
+                    color: "white"
                     visible: persistentSettings.fiatPriceEnabled
                     anchors.right: parent.right
                     anchors.rightMargin: 45
                     anchors.top: parent.top
-                    anchors.topMargin: 29
+                    anchors.topMargin: 28
+                    themeTransition: false
 
                     MouseArea{
                         anchors.fill: parent
diff --git a/main.qml b/main.qml
index 36ddf6d..992999f 100644
--- a/main.qml
+++ b/main.qml
@@ -1125,6 +1125,7 @@ ApplicationWindow {
         rootItem.state = "wizard"
         // reset balance
         leftPanel.balanceText = leftPanel.unlockedBalanceText = walletManager.displayAmount(0);
+        fiatApiUpdateBalance(0, 0);
         // disable timers
         userInActivityTimer.running = false;
         simpleModeConnectionTimer.running = false;
```

xmrdsc commented:
@selsta Thanks. That cleans up the balance between switching wallets.

@mmbyday That reminds me, we have similar problems here:

https://github.com/monero-project/monero-gui/blob/97cd215491a44c753ee1450712b3b8d93efdcc37/main.qml#L105-L108

Problem is related to openssl + buildbot is not on 5.9.7 yet for Windows. I think we should wait for that first.

luigi1111 commented:
@xmrdsc please rebase.

xmrdsc commented:
Rebased
